apis/v1alpha1/gateway_types.go

```diff
@@ -263,6 +263,16 @@ type Listener struct {
 	// is required if the Protocol field is "HTTPS" or "TLS" and
 	// ignored otherwise.
 	//
+	// The association of SNIs to Certificate defined in TLSConfig is
+	// defined based on the value of HostnameMatchType for this listener:
+	// - "Domain": Certificate should be used for the domain and its
+	//   first-level subdomains.
+	// - "Exact": Certificate should be used for the domain only.
+	// - "Any": Certificate in TLSConfig is the default certificate to use.
+	//
+	// The GatewayClass MUST use the longest matching SNI out of all
+	// available certificate for any TLS handshake.
```

Comment:
I'm not sure in general about this paragraph. The way I had thought about this was that (in this PR) there is a single CertificateRef, and that contains a single server certificate. The `Hostname` field defines how to match the requested SNI servername, so the proxy should just offer up the certificate it has and hope for the best.

This paragraph seems like it is written to capture the case where the proxy may have multiple certificates to choose from. However, if that is the case, I'd prefer to simply specify that the proxy should serve the certificate that is the most specific match (e.g. it's better to match the SNI name exactly than to serve a wildcard certificate).

Reply:
> This paragraph

Do you mean the stanza above this which defines how HostnameMatchType and certificates relate or specifically for this line?

Reply:
Yup. Do we need to specify name matching rules for certificates, or is it enough to just say "serve the certificate resolved by CertificateRef`? Maybe that depends on the contents of that resource?

Reply:
The reason I think we should define it is because with compatible listeners, we get into muddy waters here.
We can skip defining the behavior if we do not want to enforce the behavior at all.

Reply:
Ping @jpeach.

Reply:
Let's just merge and see how it goes.